### PR TITLE
fix(cacheServer): fix init cache server client

### PR DESCRIPTION
**Pull request by Harasz**

### CODE REVIEW

File: lib/LoginStrategy.ts [hunk 1]
```diff
@@ -43,14 +43,14 @@ export class LoginStrategy extends BaseStrategy {
...
   private readonly provider: providers.JsonRpcProvider
-  private readonly cacheServerClient!: CacheServerClient
   private readonly numberOfBlocksBack: number
   private readonly ensResolver: PublicResolver
   private readonly didResolver: Resolver
   private readonly ipfsStore: DidStore
   private readonly acceptedRoles: Set<string>
   private readonly strategyAddress?: string
   private readonly privateKey: string
+  private cacheServerClient?: CacheServerClient
```
Comment (jrhender) [inline lib/LoginStrategy.ts:53]: @Harasz So is this the key part of the change? Before we had `cacheServerClient` as an asserted not-undefined and not it can be undefined?

Comment (Harasz) [reply lib/LoginStrategy.ts:53]: The key part is that we are set a reference to `cacheServerClient` after login to cache server not before. That change was just to make types more predictable.

File: lib/LoginStrategy.ts [hunk 2]
```diff
@@ -105,6 +100,17 @@ export class LoginStrategy extends BaseStrategy {
...
   }
+
+  private async initCacheServer(privateKey: string, cacheServerUrl: string) {
+    const cacheServerClientInstance = new CacheServerClient({
+      privateKey,
+      provider: this.provider,
+      url: cacheServerUrl,
+    });
+    await cacheServerClientInstance.login();
+    this.cacheServerClient = cacheServerClientInstance;
```
Comment (Harasz) [inline lib/LoginStrategy.ts:111]: @jrhender The key part

Comment (jrhender) [reply lib/LoginStrategy.ts:111]: Ahhhh, I see. The key is the assignment of `this.cacheServerClient` after `login()` is completed.

File: lib/BaseStrategy.ts [hunk 3]
```diff
@@ -74,7 +74,7 @@ export abstract class BaseStrategy extends Strategy {
         return this.error(err)
       }
       if (!user) {
-        return self.fail(info)
+        return this.fail(info)
```
Comment (jrhender) [inline lib/BaseStrategy.ts:77]: Why this change? Was this just a bug that you noticed?

**Review by jrhender — COMMENTED** [review]
Another question, but I'm understanding better now

File: lib/LoginStrategy.ts (hunk 2 above)
Comment (Harasz) [reply lib/LoginStrategy.ts:111]: Yeah, that is the point

File: lib/BaseStrategy.ts (hunk 3 above)
Comment (Harasz) [reply lib/BaseStrategy.ts:77]: Idk, why typescript did not report that, but there is not `self` variable. While I was debugging I once run into that if statement and it throws me runtime error. It's look like someone was thinking about coding in python ;)

**Review by jrhender — APPROVED** [review]
Looks good to me
